### DISCUSSION TITLE
Suggestion for forking your work to use on an OpenSource EBike but custom data / ANT channel, not ANT+ LEV

### DISCUSSION THREAD

**Discussion by casainho** [maca88/E-Bike-Edge-MultiField#4] [General]
Hi.

Many thanks for sharing all this work as OpenSource!!

I am riding bicycles since always and with EBikes, even more, it has been a big part of my life and will continue to be - that is why I also keep developing technology and knowledge as OpenSource for EBikes.

We have our won DIY wireless board to make the EBike motors with ANT and Bluetooth connections and we have ANT+ LEV implemented on the firmware. Recently I started to test ANT generic channel to communicate with Garmin fields and I was thinking in having on the Edge the max of 10 fields that as you may know, the way Garmin says, must be a ANT channel for each unique field. The problem I ended up to have is on the Garmin side, where to much channels / fields make each one to not being able to connect - using 2 or 3 fields seems to be ok, but not much more then that. This while keeping their ANT+ LEV native implementation that always works well. We also have a fully wireless EBike remote for ANT+ LEV (with wired connection to brake sensors), to control our EBike motor plus ANT+ Controls on the same remote to control page change on the Edge, so, the our wireless remote do both ANT+ LEV and ANT+ Controls.

I do not like the idea to have only one big field as you did but I do not think there is other option currently as the ANT and Garmin fields work, as you say, if the fields could share data, that would be the solution for this issue with to many ANT channels...

So, I am thinking if I should fork and reuse your code, but will not be for ANT+ LEV but for the all motor data like motor current, battery current, motor temperature in degrees celsius, motor PWM, etc, etc. We have the same data as ANT+ LEV but even more, so, my idea is to make something specific.

Do you have any suggestions? Thank you.

Our project page: https://opensourceebike.github.io/

Our EBike motor wireless board (ANT and Bluetooth): 
![image](https://user-images.githubusercontent.com/147128/112225450-b9de5880-8c24-11eb-8239-48f540d87dd1.png)

![image](https://user-images.githubusercontent.com/147128/112225768-32ddb000-8c25-11eb-8e6e-a2b51d847c0f.png)

Our fully wireless EBike remote, with wired connection to brake sensors (ANT+ LEV and ANT+ Controls):
![image](https://user-images.githubusercontent.com/147128/112225626-f8741300-8c24-11eb-8abb-409f7ea81f04.png)

![image](https://user-images.githubusercontent.com/147128/112225654-032ea800-8c25-11eb-97e8-4057f614bcd8.png)

![image](https://user-images.githubusercontent.com/147128/112225672-0c1f7980-8c25-11eb-80d8-bf9d917a6b68.png)

![image](https://user-images.githubusercontent.com/147128/112225700-16da0e80-8c25-11eb-958d-981d662ab3d4.png)

**Comment by maca88** [comment]
Hi

I did check the project as it [was mentioned](https://github.com/maca88/E-Bike-Edge-MultiField/issues/1#issuecomment-774090631) in one of the issues and I must say, I am really impresseed what you all managed to achieve :)


> I do not like the idea to have only one big field as you did but I do not think there is other option currently as the ANT and Garmin fields work

I don't like it either, but as you realized it is the only reliable way to show more than one LEV field. It is really strange why Garmin didn't made the LEV profile available in their [AntPlus module](https://developer.garmin.com/connect-iq/api-docs/Toybox/AntPlus.html), where we could use the [onMessage](https://developer.garmin.com/connect-iq/api-docs/Toybox/AntPlus/DeviceListener.html#onMessage-instance_function) callback to handle the data without needing to open additional ANT channels.

> So, I am thinking if I should fork and reuse your code, but will not be for ANT+ LEV but for the all motor data like motor current, battery current

Sure, one option is to extend the ANT+ LEV profile by adding custom data pages where you can have the additional information that do not exist in the ANT+ LEV profile. In this way a generic ANT+ LEV display can still be used in case the user does not own a Garmin device.

> Do you have any suggestions?

The first thing that I would like to point out is memory. Low end devices like Edge 130 have only 32KB memory available for a data field application, where 28.6KB can actually be used by the application. In the current state of this data field, there is aprox. 5KB free memory left, so you may need to remove some functionality in case it won't be enough to implement the additional data that you would like to display. This data field dynamicaly calculates the field positions based on its placement and the selected layout, which has its pros and cons:
Pros:
+ Can be used alongside native fields by displaying up to four LEV fields

Cons:
- Complex field position calculation logic that requires more memory
- The calculation logic supports only Edge devices (rectangular screens)

One way to reduce memory and simplify the code is to use the field positions data inside the devices.xml file ([check here](https://forums.garmin.com/developer/connect-iq/f/discussion/256346/handling-differences-between-system-fonts-in-simulators-and-devices/1224191#1224191) for more info) instead of having the dynamic calculation logic, but at a cost of having only a full screen support:
Pros:
+ Simpler code and less memory usage
+ Can be used for all devices

Cons:
- Cannot be used alongside native fields

Another thing to point out is that the data field is "limited" to 16 fields as it uses 8 byte numbers to simulate a dictionary:
https://github.com/maca88/E-Bike-Edge-MultiField/blob/99591a2ff1efa041492c6ba3be18a43ad5500674/Source/EBikeEdgeMultiField/source/EBikeMultiField.mc#L398-L401

https://github.com/maca88/E-Bike-Edge-MultiField/blob/99591a2ff1efa041492c6ba3be18a43ad5500674/Source/EBikeEdgeMultiField/source/EBikeMultiField.mc#L243

although Monkey C has dictionaries, they are very memory expensive.

So my suggestion would be to fork the repository, then start extending the `EBikeSensor.mc` file by adding the additional custom data pages and then see how much memory would end up consuming. Based on the result, you could then decide whether to leave the field positioning logic as it is or to find an alternative.

**Comment by casainho** [comment]
Thanks for the extensive help!!

@rananna see this message please, to know how the development may go.

**Comment by casainho** [comment]
I want to have lower work possible. If I understand correctly, I can only have 16 different fields / variables but 13 are already used by ANT+ LEV, so I can only add 3. I need like 14. So maybe the only solution is to override the current 13 fields.
I need to keep LEV working on the firmware and I would add an extra ANT channel that would work only for my field, so user could keep using your field and my field simultaneous.

**Comment by maca88** [comment]
> If I understand correctly, I can only have 16 different fields

Yes, but it could be modified by using an extra 8 byte number for additional 16 fields.

>  I would add an extra ANT channel that would work only for my field

Having an additional ANT channel for providing custom data would be even better and would lower the amount of work that need to be done as the memory wouldn't be an issue anymore. The additional channel should be opened with a different network key in order to prevent any collisions. In this way, the user could also use this data field and the forked one at the same time and be able to display up to two fields from this and two fields from the forked data field alongside with 8 native fields. So yes, I think this would be the best and easiest solution overall.
